### PR TITLE
Rewritten 8051 disassembler

### PR DESCRIPTION
**Pull request by astuder**
As reported in issue #7737 the 8051 disassembler had many issues, so I rewrote it from scratch. 

- added missing op codes
- fixed incomplete/broken disassembly
- fixed address decoding for branch instructions
- added # prefix for immediate values
- adapted 8051 anal plugin as it had dependency on old disasm, but did not change ESIL part
- updated and validated regression tests in t.asm/8051 (100% OK)

### CODE REVIEW

File: libr/anal/p/anal_8051.c [hunk 1]
```diff
@@ -486,79 +486,134 @@ static int set_reg_profile(RAnal *anal) {
 	return r_reg_set_profile_string (anal->reg, p);
 }
 
-
-// TODO: Cleanup the code, remove unneeded data copies
+static inline ut16 arg_offset (ut16 pc, ut8 offset) {
+	if (offset < 0x80) {
+		return pc + offset;
+	} else {
```
Comment (radare) [inline libr/anal/p/anal_8051.c]: dont else after return

File: libr/asm/p/asm_8051.c [hunk 2]
```diff
@@ -11,28 +11,10 @@
...
-		}
-		free (tmp);
-	}
-	if (!*op->buf_asm) {
-		op->size = 1;
-		return -1;
-	}
-	return (op->size = o.length);
+	int dlen = _8051_disas (a->pc, op, buf, len);
+	if (dlen < 0) dlen = 0;
```
Comment (radare) [inline libr/asm/p/asm_8051.c]: missing braces {}

File: libr/asm/arch/8051/8051_disas.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		return pc - offset;
+	}
+}
+
+int _8051_disas (ut64 pc, RAsmOp *op, const ut8 *buf, ut64 len) {
+	int i = 0;
+	while (_8051_ops[i].string != NULL && _8051_ops[i].op != (buf[0] & ~_8051_ops[i].mask))
+		i++;
+
+	if (_8051_ops[i].string != NULL) {
```
Comment (radare) [inline libr/asm/arch/8051/8051_disas.c]: != NULL is not necessary

File: libr/asm/arch/8051/8051_disas.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		return pc + offset;
+	} else {
+		offset = 0 - offset;
+		return pc - offset;
+	}
+}
+
+int _8051_disas (ut64 pc, RAsmOp *op, const ut8 *buf, ut64 len) {
+	int i = 0;
+	while (_8051_ops[i].string != NULL && _8051_ops[i].op != (buf[0] & ~_8051_ops[i].mask))
```
Comment (radare) [inline libr/asm/arch/8051/8051_disas.c]: missing braces

File: libr/asm/arch/8051/8051_disas.c [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#include <r_lib.h>
+#include <string.h>
+
+#include "8051_ops.h"
+#include "8051_disas.h"
+
+static inline ut16 arg_offset (ut16 pc, ut8 offset) {
+	if (offset < 0x80) {
+		return pc + offset;
+	} else {
```
Comment (radare) [inline libr/asm/arch/8051/8051_disas.c]: dont else after return

File: libr/asm/arch/8051/8051_disas.c [hunk 6]
```diff
@@ -0,0 +1,101 @@
+/* radare - LGPL - Copyright 2015-2016 - pancake, condret, riq, qnix */
```
Comment (radare) [inline libr/asm/arch/8051/8051_disas.c]: wrong copyright. update year and add your name

File: libr/asm/arch/8051/8051.c [hunk 7]
```diff
@@ -1,267 +0,0 @@
-#if 0
```
Comment (radare) [inline libr/asm/arch/8051/8051.c]: poor file deleted :_ we will miss him

File: libr/anal/p/anal_8051.c [hunk 8]
```diff
@@ -486,79 +486,134 @@ static int set_reg_profile(RAnal *anal) {
...
-		if (buf_asm[0]=='j' || (buf_asm[0] && buf_asm[1] == 'j')) {
-			op->type = R_ANAL_OP_TYPE_JMP;
-			if (o.operand == OFFSET) {
-				op->jump = o.addr + addr + o.length;
-			} else {
-				op->jump = o.addr;
-			}
-			op->fail = addr + o.length;
+		if (_8051_ops[i].arg1 == A_ADDR11) {
+			op->jump = ((buf[0]&0xe0)<<3) + buf[1];
```
Comment (radare) [inline libr/anal/p/anal_8051.c]: spaces around << & * ..

File: libr/asm/arch/8051/8051_disas.c (hunk 3 above)
Comment (astuder) [inline libr/asm/arch/8051/8051_disas.c]: The while will stop when it finds the op. string is only NULL for invalid op.

**Comment by radare** [comment]
no. i mean, that if (_8051_ops[i].mask) { is enough, no need to put the != NULL

**Pull request merged**